File: packages/backend/src/expose/records.ts
import type { ControlService, DeviceRecord, ExposeRecord } from '../storage/control'
import { errorMessage } from '@demicodes/utils'

/** One expose lifetime (`expose.md` § Lifetime): one hour, renewable forever. */
export const EXPOSE_LIFETIME_MS = 60 * 60_000

/** Concurrent relayed connections per expose; the 65th answers 503. */
export const EXPOSE_CONNECTION_LIMIT = 64

/** The expiry sweep's production cadence (`expose.md` § Lifetime). */
const DEFAULT_SWEEP_MS = 60_000

const BASE32_ALPHABET = 'abcdefghijklmnopqrstuvwxyz234567'

/** 128 random bits as 26 lowercase base32 characters — a DNS label and the only credential. */
export function generateExposeId(): string {
  const bytes = crypto.getRandomValues(new Uint8Array(16))
  let bits = 0
  let acc = 0
  let out = ''
  for (const byte of bytes) {
    acc = (acc << 8) | byte
    bits += 8
    while (bits >= 5) {
      bits -= 5
      out += BASE32_ALPHABET[(acc >> bits) & 31]
    }
  }
  // The final partial character carries the leftover bits: 128 bits is 26
  // base32 characters, not the 25 the whole groups spell.
  if (bits > 0)
    out += BASE32_ALPHABET[(acc << (5 - bits)) & 31]
  return out
}

/**
 * A bare port means `127.0.0.1:<port>`; anything else is used exactly as
 * given — the device's network is the boundary, not Demi.
 */
export function normalizeExposeAddress(address: string): string {
  return /^\d+$/.test(address) ? `127.0.0.1:${address}` : address
}

/** Split a stored address into the runner's `net_open` host and port. */
export function exposeTarget(address: string): { host: string; port: number } {
  const index = address.lastIndexOf(':')
  if (index === -1)
    throw new Error(`expose address ${address} has no port`)
  const port = Number(address.slice(index + 1))
  if (!Number.isInteger(port) || port < 1 || port > 65535)
    throw new Error(`expose address ${address} has no valid port`)
  return { host: address.slice(0, index), port }
}

/** The error codes `expose.md` § Commands and `web-api.md` § Exposes define. */
export class ExposeError extends Error {
  constructor(
    readonly code: 'expose_unavailable' | 'device_offline' | 'expose_not_found' | 'device_not_found',
    message: string
  ) {
    super(message)
    this.name = 'ExposeError'
  }
}

/** The `host:port` of the service and the hostname that reaches it. */
export function exposeHostname(id: string, domain: string): string {
  return `${id}.${domain}`
}

/** `expose.md` § Deployment: the scheme of the printed URLs follows the forwarded protocol. */
export function exposeUrl(
  id: string,
  domain: string,
  scheme: string
): string {
  return `${scheme}://${exposeHostname(id, domain)}/`
}

/** The response shape of an expose on every surface (`web-api.md` § Exposes). */
export interface ExposeView {
  id: string
  deviceId: string
  address: string
  url: string
  createdAt: string
  expiresAt: string
}

/**
 * The expose records and their one-hour lifetime (`expose.md`). Owns the
 * clock, the expiry sweep and the registry of live relayed connections: every
 * path that destroys a record — expiry, remove, Cloud stop, revocation —
 * ends its connections through here.
 */
export class Exposes {
  private readonly sweepTimer: ReturnType<typeof setInterval> | null = null
  private readonly connections = new Map<string, Set<() => void>>()

  constructor(private readonly deps: {
    control: ControlService
    /** `DEMI_EXPOSE_DOMAIN`; null disables the whole feature. */
    domain: string | null
    /** The registry's live-connection check: creation requires a connected device. */
    deviceOnline: (deviceId: string) => boolean
    now: () => number
    /** The fallback scheme of printed URLs when no request supplies one. */
    scheme: () => string
    /** Tests raise the sweep cadence; production sweeps every 60 s. */
    sweepMs?: number
    log?: (line: string) => void
  }) {
    this.sweepTimer = setInterval(
      () => void this.sweep().catch(error =>
        this.deps.log?.(`expose sweep: ${errorMessage(error)}`)
      ),
      deps.sweepMs ?? DEFAULT_SWEEP_MS
    )
    // The sweep is housekeeping: it must not hold the process open.
    this.sweepTimer.unref()
  }

  get domain(): string | null {
    return this.deps.domain
  }

  close(): void {
    if (this.sweepTimer !== null)
      clearInterval(this.sweepTimer)
    for (const close of [...this.connections.values()].flatMap(set => [...set]))
      close()
    this.connections.clear()
  }

  /** Creation: one hour from now, on a connected caller-owned device. */
  async add(
    userId: string,
    device: DeviceRecord,
    address: string
  ): Promise<ExposeRecord> {
    if (this.deps.domain === null)
      throw new ExposeError(
        'expose_unavailable',
        'This backend has no expose domain configured (DEMI_EXPOSE_DOMAIN)'
      )
    if (!this.deps.deviceOnline(device.id))
      throw new ExposeError(
        'device_offline',
        `The device ${device.id} is offline; connect it before exposing a service`
      )
    const now = this.deps.now()
    return this.deps.control.createExpose({
      id: generateExposeId(),
      userId,
      deviceId: device.id,
      address: normalizeExposeAddress(address),
      createdAt: new Date(now).toISOString(),
      expiresAt: new Date(now + EXPOSE_LIFETIME_MS).toISOString(),
    })
  }

  /** The user's live exposes, soonest expiry first; expired rows die here too. */
  async list(userId: string): Promise<ExposeRecord[]> {
    await this.sweep()
    return (await this.deps.control.listExposes(userId))
      .filter(record => !this.expired(record))
  }

  async renew(userId: string, id: string): Promise<ExposeRecord> {
    await this.sweep()
    const record = await this.deps.control.getExpose(id)
    if (!record || record.userId !== userId || this.expired(record))
      throw new ExposeError('expose_not_found', `No expose ${id}`)
    const expiresAt = new Date(
      this.deps.now() + EXPOSE_LIFETIME_MS
    ).toISOString()
    return (await this.deps.control.renewExpose(id, userId, expiresAt))!
  }

  async remove(userId: string, id: string): Promise<void> {
    const record = await this.deps.control.getExpose(id)
    if (!record || record.userId !== userId || this.expired(record))
      throw new ExposeError('expose_not_found', `No expose ${id}`)
    await this.destroy(id)
  }

  /**
   * The record behind a hostname, for the relay only: null when unknown, and
   * destroyed-and-null once expired.
   */
  async liveRecord(id: string): Promise<ExposeRecord | null> {
    const record = await this.deps.control.getExpose(id)
    if (!record)
      return null
    if (this.expired(record)) {
      await this.destroy(id)
      return null
    }
    return record
  }

  /** Deletes expired rows and ends their connections. */
  async sweep(): Promise<void> {
    const destroyed = await this.deps.control.deleteExpiredExposes(
      new Date(this.deps.now()).toISOString()
    )
    for (const id of destroyed)
      this.endConnections(id)
  }

  /** Every expose on a device: a Cloud stop and device revocation. */
  async destroyForDevice(deviceId: string): Promise<void> {
    for (const id of await this.deps.control.deleteExposesByDevice(deviceId))
      this.endConnections(id)
  }

  /** Destroys one record at once; its URL stops working. */
  private async destroy(id: string): Promise<void> {
    await this.deps.control.deleteExpose(id)
    this.endConnections(id)
  }

  private expired(record: ExposeRecord): boolean {
    return Date.parse(record.expiresAt) <= this.deps.now()
  }

  url(record: ExposeRecord, scheme?: string): string {
    return exposeUrl(
      record.id,
      this.deps.domain!,
      scheme ?? this.deps.scheme()
    )
  }

  /** The response shape every surface prints (`web-api.md` § Exposes). */
  view(record: ExposeRecord): ExposeView {
    return {
      id: record.id,
      deviceId: record.deviceId,
      address: record.address,
      url: this.url(record),
      createdAt: record.createdAt,
      expiresAt: record.expiresAt,
    }
  }

  /**
   * Registers one relayed connection on an expose; the closer ends it from
   * either side. Returns null at the concurrent-connection limit.
   */
  begin(id: string, close: () => void): (() => void) | null {
    let live = this.connections.get(id)
    if (!live) {
      live = new Set()
      this.connections.set(id, live)
    }
    if (live.size >= EXPOSE_CONNECTION_LIMIT)
      return null
    live.add(close)
    let released = false
    return () => {
      if (released)
        return
      released = true
      live!.delete(close)
      if (live!.size === 0)
        this.connections.delete(id)
    }
  }

  private endConnections(id: string): void {
    const live = this.connections.get(id)
    if (!live)
      return
    this.connections.delete(id)
    for (const close of [...live])
      close()
  }
}

File: crates/backend/src/expose/records.rs
//! The expose records in the user's shard (`expose.md` § The expose record,
//! § Lifetime; `web-api.md` § Exposes): creation on a connected device, the
//! owner's list, renewal and removal, which the routes and the commands
//! share, and the destruction that ends an expose's connections. Every
//! surface shows an expose with its URL, whose scheme and port are the
//! backend's public URL's. Without an expose domain the instance has no
//! exposes: creation is refused as unavailable, and there is none to list,
//! renew or remove.

use demi_web_api::devices::DeviceKind;
use demi_web_api::error::ErrorCode;
use demi_web_api::exposes::{ExposeAddress, ExposeDto};
use demi_web_api::ids::{DeviceId, ExposeId};
use jiff::SignedDuration;
use url::Url;

use super::ExposeDomain;
use crate::shard::Shard;
use crate::storage::StorageError;
use crate::storage::exposes::ExposeRecord;

/// How long an expose lives from its creation or its last renewal.
pub(super) const LIFETIME: SignedDuration = SignedDuration::from_hours(1);

/// Why an expose operation was refused.
#[derive(Debug, thiserror::Error)]
pub(crate) enum ExposeError {
    /// The instance has no expose domain.
    #[error("This backend has no expose domain configured (DEMI_EXPOSE_DOMAIN)")]
    Unavailable,
    /// The device is not the caller's.
    #[error("No such device")]
    DeviceNotFound,
    /// The device's runner is not connected, or the Cloud is not running.
    #[error("The device {0} is offline; connect it before exposing a service")]
    DeviceOffline(DeviceId),
    /// The caller has no expose of this id, or it expired.
    #[error("No expose {0}")]
    NotFound(String),
    #[error(transparent)]
    Storage(#[from] StorageError),
}

impl ExposeError {
    /// The code and HTTP status a refusal answers with (`web-api.md`
    /// § Exposes); none for a failure the caller could not cause.
    pub(crate) fn code(&self) -> Option<(ErrorCode, u16)> {
        match self {
            Self::Unavailable => Some((ErrorCode::ExposeUnavailable, 409)),
            Self::DeviceNotFound => Some((ErrorCode::DeviceNotFound, 404)),
            Self::DeviceOffline(_) => Some((ErrorCode::DeviceOffline, 409)),
            Self::NotFound(_) => Some((ErrorCode::ExposeNotFound, 404)),
            Self::Storage(_) => None,
        }
    }
}

/// A new expose's id: 128 random bits as 26 lowercase base32 characters.
fn new_expose_id() -> ExposeId {
    let bits: [u8; 16] = rand::random();
    let text = data_encoding::BASE32_NOPAD.encode(&bits).to_ascii_lowercase();
    ExposeId::try_from(text).expect("128 bits are 26 base32 characters")
}

/// The public URL of the expose `id`: its hostname under `domain`, with the
/// scheme of `backend` and its port unless that is the scheme's default.
pub(crate) fn expose_url(id: &ExposeId, domain: &ExposeDomain, backend: &Url) -> String {
    let port = backend.port().map(|port| format!(":{port}")).unwrap_or_default();
    format!("{}://{id}.{}{port}/", backend.scheme(), domain.as_str())
}

impl Shard {
    /// A new expose of `address` on the user's device `device`, for an
    /// hour: the device is connected, and a Cloud is running.
    pub(crate) async fn add_expose(&self, device: &DeviceId, address: ExposeAddress) -> Result<ExposeDto, ExposeError> {
        let services = self.services();
        let record = services
            .control
            .device(device.clone())
            .await?
            .filter(|record| record.user == *self.user())
            .ok_or(ExposeError::DeviceNotFound)?;
        let domain = services.expose_domain.as_ref().ok_or(ExposeError::Unavailable)?;
        let connected = match record.kind {
            DeviceKind::User => self.devices().online(device),
            DeviceKind::Managed => self.devices().online(device) && self.cloud().runs(device),
        };
        if !connected {
            return Err(ExposeError::DeviceOffline(device.clone()));
        }
        // The database takes the record in the step that checked the
        // device: a Cloud that stops after the check destroys it with the
        // Cloud's other exposes.
        let created = services
            .control
            .create_expose(new_expose_id(), self.user().clone(), device.clone(), address, LIFETIME)
            .await?;
        Ok(self.expose_dto(created, domain))
    }

    /// The user's live exposes, soonest expiry first; the expired ones are
    /// destroyed.
    pub(crate) async fn list_exposes(&self) -> Result<Vec<ExposeDto>, StorageError> {
        let Some(domain) = &self.services().expose_domain else {
            return Ok(Vec::new());
        };
        let listed = self.services().control.user_exposes(self.user().clone()).await?;
        self.exposes().end(&listed.expired);
        Ok(listed
            .live
            .into_iter()
            .map(|record| self.expose_dto(record, domain))
            .collect())
    }

    /// Moves the expiry of the user's expose `id` to an hour from now.
    pub(crate) async fn renew_expose(&self, id: &ExposeId) -> Result<ExposeDto, ExposeError> {
        let not_found = || ExposeError::NotFound(id.to_string());
        let domain = self.services().expose_domain.as_ref().ok_or_else(not_found)?;
        let record = self.owned_expose(id).await?.ok_or_else(not_found)?;
        if self.destroy_if_expired(&record).await? {
            return Err(not_found());
        }
        let renewed = self
            .services()
            .control
            .renew_expose(id.clone(), self.user().clone(), LIFETIME)
            .await?
            .ok_or_else(not_found)?;
        Ok(self.expose_dto(renewed, domain))
    }

    /// Destroys the user's expose `id` at once. One that had expired is
    /// destroyed too, and answers as not found.
    pub(crate) async fn remove_expose(&self, id: &ExposeId) -> Result<(), ExposeError> {
        let not_found = || ExposeError::NotFound(id.to_string());
        self.services().expose_domain.as_ref().ok_or_else(not_found)?;
        let record = self.owned_expose(id).await?.ok_or_else(not_found)?;
        if self.destroy_if_expired(&record).await? {
            return Err(not_found());
        }
        self.destroy_expose(id).await?;
        Ok(())
    }

    /// The user's expose `id`, expired or not.
    pub(super) async fn owned_expose(&self, id: &ExposeId) -> Result<Option<ExposeRecord>, StorageError> {
        let record = self.services().control.expose(id.clone()).await?;
        Ok(record.filter(|record| record.user == *self.user()))
    }

    /// Destroys the user's `record` if it expired (`expose.md` § Lifetime),
    /// as every read of one does; answers whether it had.
    pub(super) async fn destroy_if_expired(&self, record: &ExposeRecord) -> Result<bool, StorageError> {
        if self.services().clock.now() < record.expires_at {
            return Ok(false);
        }
        self.destroy_expose(&record.id).await?;
        Ok(true)
    }

    /// Destroys the user's expose `id`: its record goes, and its
    /// connections end.
    async fn destroy_expose(&self, id: &ExposeId) -> Result<(), StorageError> {
        self.services().control.delete_expose(id.clone()).await?;
        self.exposes().end([id]);
        Ok(())
    }

    /// Destroys every expose on the user's `device`, as the device's
    /// revocation and a Cloud's stop do. A failure is logged: the device's
    /// runner, which goes next, ends the connections all the same, and the
    /// records expire within the hour.
    pub(crate) async fn destroy_exposes_on(&self, device: &DeviceId) {
        match self.services().control.delete_device_exposes(device.clone()).await {
            Ok(ids) => self.exposes().end(&ids),
            Err(error) => tracing::error!(device = %device, "the exposes of a device could not be destroyed: {error}"),
        }
    }

    /// The expose as every surface shows it.
    fn expose_dto(&self, record: ExposeRecord, domain: &ExposeDomain) -> ExposeDto {
        let backend = self
            .services()
            .public_url
            .get()
            .expect("the backend listens before it serves a request or a command");
        ExposeDto {
            url: expose_url(&record.id, domain, backend.url()),
            id: record.id,
            device_id: record.device,
            address: record.address,
            created_at: record.created_at,
            expires_at: record.expires_at,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn an_expose_id_is_a_dns_label_of_128_random_bits() {
        let id = new_expose_id();
        assert_eq!(id.as_str().len(), 26);
        assert_ne!(id, new_expose_id());
    }

    #[test]
    fn an_expose_url_takes_the_scheme_and_any_other_than_the_default_port_from_the_backend_url() {
        let id = ExposeId::try_from("k7x2maqw4p3s6tavaw2y4z6aab").unwrap();
        let local: ExposeDomain = "expose.localhost".parse().unwrap();
        let backend = Url::parse("http://localhost:3271").unwrap();
        assert_eq!(expose_url(&id, &local, &backend), format!("http://{id}.expose.localhost:3271/"));
        let public: ExposeDomain = "expose.demi.example".parse().unwrap();
        for backend in ["https://demi.example", "https://demi.example:443/", "https://demi.example/api"] {
            let backend = Url::parse(backend).unwrap();
            assert_eq!(expose_url(&id, &public, &backend), format!("https://{id}.expose.demi.example/"));
        }
    }
}
